document-store/src/main/java/org/hypertrace/core/documentstore/expression/AggregateExpression.java

```diff
@@ -0,0 +1,16 @@
+package org.hypertrace.core.documentstore.expression;
+
+import lombok.Value;
+
+/**
+ * Expression representing aggregation in a query.
+ *
+ * <p>Example: SUM(col1) can be constructed as <code>
+ *    AggregatorExpression.of(Aggregator.SUM, LiteralExpression.of("col1"));
+ * </code>
+ */
+@Value(staticConstructor = "of")
+public class AggregateExpression implements Projectable, Sortable {
```

Comment:
Marker interfaces are often considered an antipattern (here's the first article I found talking about why: https://blog.ndepend.com/marker-interface-isnt-pattern-good-idea/ )- it seems like what we're really going for here is trying to get around the fact that java doesn't have a strong concept of unions. Either putting methods on the interfaces or using a different pattern (such as how oneof is represented in the java generated by protoc) would likely be a better fit.

Reply:
Is it possible for something to be projectable but not sortable? Or vice versa? If not, it may make sense to merge the two.

Reply:
This is very useful! Thank you. :) 
Read this article https://hackernoon.com/are-interfaces-code-smell-bd19abc266d3 and changed it to a visitor design pattern. The only difference here is that we have multiple hierarchies here instead of one.


On another note, ConstantExpressions are projectable and not sortable though SQL allows integers in "ORDER BY" clause. For example, ORDER BY 2, sorts by 2nd field in the projection list. We do not support that as of today as this implementation is not going to be tied to any specific SQL.